### ISSUE TITLE
accessibility, accordion: screen reader leest afgerond / niet afgerond niet wegens decoratieve icoontjes

### ISSUE THREAD

**Issue by timveld** [dso-toolkit/dso-toolkit#309]
Betreft [accordion](https://www.dso-toolkit.nl/7.1.0/components/detail/accordion.html) 

De afgeronde secties (zoals woning) hebben klasse "dso-success"  en de nog niet afgeronde secties ("plaats van het bouwwerk")  hebben klasse "dso-danger". 
In de CSS is kennelijk gespecificeerd dat "DSO-success" groen is en voorafgegaan door een decoratief plaatje (succesvinkje); "dso-danger" is rood en wordt voorafgegaan door een omgekeerd vinkje. 
Visueel is dat goed: je geeft zowel met kleur als met een alternatief (de vinkjes) aan of betreffende secties compleet zijn ingevuld. 
Echter met een screen reader is er geen enkele manier om te weten of een sectie succesvol is ingevuld; want het vinkje wordt genegeerd (decoratieve afbeelding) en de kleur wordt niet standaard voorgelezen. 

Na wat onderzoek is mijn advies de vinkjes niet als decoratieve afbeeldingen in CSS op te nemen (ze zijn M.I. ook niet decoratief maar hebben wel degelijk een functie), je kunt ze dan als gewone img met een goede alt tag in de HTML opnemen. Deze alt tag moet wel kort zijn, realiseer je dat hij volledig wordt uitgesproken door de screen reader. Ik denk dat "afgerond / niet afgerond" de beste beschrijvingen voor deze vinkjes zijn. 
Dit stemt overeen met het verhaal in [WCAG failure #F3](https://www.w3.org/TR/2008/NOTE-WCAG20-TECHS-20081211/F3).

Andere mogelijkheden die ik heb overwogen: 
- De informatie ook tekstueel weergeven, dan krijg je "woning (afgerond) of "woning (niet afgerond) op het scherm, dan is het M.I. juist om zowel kleur als het vinkje decoratief te houden. Misschien zou dat voor bepaalde gebruikersgroepen nog prettiger werken, maar neemt wel veel ruimte.
- Gebruik van Aria om dit over te brengen (stuit in ieder geval op het bezwaar dat deze oplossing volledig op Aria leunt terwijl Aria nog niet overal goed werkt). 
- Tekst "afgerond / niet afgerond" via een 'screen-reader-only' class via CSS visueel verbergen (M.I. veel te instabiel en hier niet nodig). 

Er is een Vergelijkbaar probleem bij bijvoorbeeld [de alert icoontjes](https://www.dso-toolkit.nl/7.1.0/components/detail/alert.html)

**Comment by PieterVerheij** [comment]
Dit heeft wel effect op de wijzigingen gemaakt in #261.

**Comment by MichielDHVD** [comment]
Het gaat er dus om dat wat er gecommuniceerd wordt ook beschikbaar is voor screenreaders. In het issue 261 gaat het er nu over of we beter <span> of <i> kunnen gebruiken. 

Het lijkt erop dat we span gaan gebruiken. Kan er geen span gebruikt worden om voor deze specifieke componenten deze icoontjes te tonen. En dan kan op de span eventueel een aria-text gebruikt worden. Of zelfs een hidden text. Misschien heeft @timveld nog een idee over? 

Is dit mogelijk?

**Comment by PieterVerheij** [comment]
@MichielDHVD Was deze vraag aan ons gesteld?

**Comment by tfrijsewijk** [comment]
"Afgerond" / "Niet afgerond" dekt de lading niet helemaal. Er zijn 4 statussen:
* Succes (`.dso-succes`)
* Waarschuwing (`.dso-warning`)
* Informatie (`.dso-info`)
* Let op (`.dso-danger`)

Deze statussen zijn er ook voor de alert.

Jouw oplossing is in het vervolg de statussen in een `<img>` met een `alt` te tonen.

Je geeft aan dat  "[..] een 'screen-reader-only' class via CSS visueel verbergen (M.I. veel te instabiel en hier niet nodig).".
Kun je dit verder toelichten?

De documentatie van Bootstrap 4 voor het Alert component schrijft het volgende over accessibility:

>**Conveying meaning to assistive technologies**
>
>Using color to add meaning only provides a visual indication, which will not be conveyed to users of assistive technologies – such as screen readers. Ensure that information denoted by the color is either obvious from the content itself (e.g. the visible text), or is included through alternative means, such as additional text hidden with the `.sr-only` class.

_https://getbootstrap.com/docs/4.3/components/alerts/#conveying-meaning-to-assistive-technologies_

Ik denk de `.sr-only` class van Bootstrap hier een ideale oplossing is:
* ✔️ Geen wijzigingen aan de voorgeschreven markup
* Of de implementator een tekst in `alt` of als inhoud van een `<span class="sr-only">` plaatst: het blijft de verantwoordelijkheid van de implementator.

Voor de accordion stel ik het volgende voor:
```html
    <div class="dso-accordion-section dso-success">
        <div class="dso-section-handle">
            <a href="#">Bouwwerkzaamheden <span class="sr-only">(Afgerond)</span></a>
        </div>
    </div>
```

```html
<div class="alert alert-warning" role="alert">
    <span class="sr-only">Waarschuwing:</span> Wijzig uw wachtwoord binnen 4 dagen om inlogproblemen te voorkomen
</div>
```

**Comment by MichielDHVD** [comment]
Goed voorstel om sr-only te gebruiken. Als dit onder de verantwoordelijkheid valt van de implementator moeten we dit goed documenteren. Als de gewenste implementatie van dit component. 

Ik tag even @timveld .

**Comment by tfrijsewijk** [comment]
Is er al nieuws?

**Comment by PieterVerheij** [comment]
@MichielDHVD @bramhoosemans @timveld 
Eens met de suggestie van @tfrijsewijk ?

**Comment by MichielDHVD** [comment]
Ik zal nog eens contact opnemen met @timveld. Probeer deze week hier een afsluitend antwoord op te geven.

**Comment by timveld** [comment]
Als Bootstrap dit goed ondersteunt is het gebruik van de .sr-only class prima en de voorgestelde oplossing is goed. .. maar.. 'dat is de verantwoordelijkheid van de implementator', bedoel je dan dat de implementator steeds zelf de spans moet toevoegen? Zelfs als je dit documenteert is het de vraag of dit goed gaat en het is dubbel werk, je zou willen dat de implementator maar op 1 plek hoeft aan te geven dat het een afgerond / waarschuwing / whatever is en dat de toolkit dan op de een of andere manier die sr-only automatisch toevoegt. Kunnen we dit morgen bij het UX overleg meenemen? Want dit punt zou op meer plaatsen kunnen gaan spelen en raakt ook aan het optimaliseren van het ontwikkelproces.

**Comment by tfrijsewijk** [comment]
>[..]  'dat is de verantwoordelijkheid van de implementator', bedoel je dan dat de implementator steeds zelf de spans moet toevoegen?

Klopt.

>Zelfs als je dit documenteert is het de vraag of dit goed gaat en het is dubbel werk, je zou willen dat de implementator maar op 1 plek hoeft aan te geven dat het een afgerond / waarschuwing / whatever ...

Mee eens

>... en dat de toolkit dan op de een of andere manier die sr-only automatisch toevoegt

Maar dat kan niet (ongeveer..).

De toolkit is alleen maar styling. Gedrag/interactie moet door de gebruiker van de toolkit worden gerealiseerd. Bij DSO wordt oa. met Angular en React gewerkt.

Het toevoegen van _de class `.sr-only`_ moet door de implementator worden gedaan. Maar er is wel een middenweg mogelijk: De eigenschappen van `.sr-only` kunnen vanuit de toolkit op een andere selector worden gezet. Dit betekent dat er vanuit de toolkit een markup voorschrift komt zonder `.sr-only`, maar nog steeds met een `span` waar de toolkit `.sr-only`-gedrag op kan stijlen.

We ontkomen niet aan de extra span: De implementator moet ergens in het DOM de status neerzetten, en we willen die status niet zichtbaar hebben: het is ondersteunende tekst voor een icoon. Het is heel gangbaar om daar een HTML element voor te gebruiken.

Heel veel winst is er met bovenstaande oplossing niet. Het "werk" moet nog steeds gedaan worden. In dit geval zou ik zelfs adviseren het niet te doen en gebruik te maken van de bouwstenen die de toolkit levert.

Maar er is nog een andere oplossing (en eigenlijk de enige oplossing): De implementator (de ontwikkelaar bij DSO die met React/Angular werkt) moet een component maken waarin wordt afgedwongen een status op de juiste manier in de markup te tonen. Het component is waar de tijd in gaat zitten.

Donderdag ben ik (waarschijnlijk? @MichielDHVD) in Rotterdam. We kunnen er dan over verder filosoferen?

**Comment by bramhoosemans** [comment]
Actie: voorbeeld met sr.only pagina op dso-toolkit.nl voor test met screenreader.

**Comment by tfrijsewijk** [comment]
Voorbeeld: https://dso-toolkit.nl/_309.status-toevoegen-wegens-decoratieve-icoontjes/components/detail/accordion.html

@timveld Wat denk je ervan?

**Comment by timveld** [comment]
1. Er staat geen spatie tussen de sr-only en de rest van de content, dus ik lees: "succes:Bouwwerkzaamheden (h2)
2. Succes / waarschuwing: beetje vage omschrijving, wordt hier niet bedoeld 'ingevuld / niet ingevuld' of 'afgerond / niet afgerond'? (als ook een evaluatie plaatsvindt moet je het vooral zo laten maar volgens mij zegt het vinkje alleen dat alle verplichte vragen zijn beantwoord)

**Comment by PieterVerheij** [comment]
@timveld 
1. ik kijk/luister met ChromeVox, daar wordt de spatie (die in de code staat) wel voorgelezen?
2. het is aan de implementator om daar een goede omschrijving te plaatsen, maar er zijn iig meer statussen dan (niet) ingevuld en (niet) afgerond.

**Comment by timveld** [comment]
1. Ik had het alleen in Firefox met NVDA gecontroleerd, maar in internet explorer gaat het wel goed en de code klopt. Je zou nog &nbsp; kunnen overwegen maar ik weet niet of dat handig is, lijkt er op dat dit een firefox-eigenaardigheid is die wij niet hoeven op te vangen. Door gebruik van de dubbele punt kan sowieso geen verwarring ontstaan. 
2. We moeten hier over toch iets documenteren. Prima om het bij de implementator te leggen want de betekenis van het vinkje kan per situatie nogal verschillen. Maar vergelijkbare situaties moeten in verschillende delen van het portal wel concistent beschreven worden en de omschrijving moet zo duidelijk mogelijk zijn. Als wordt bedoeld dat nog niet alle vragen zijn ingevuld is 'waarschuwing' een veel te onduidelijke omschrijving. Denk dat we voor de vragenbomen de keuze hebben uit 'afgerond / niet afgerond' of 'voltooid / niet voltooid', eventueel 'ingevuld / niet (/gedeeltelijk) ingevuld. Worden deze accordions ook voor andere situaties dan vragenbomen gebruikt? 
3. Wat mij betreft kan het issue worden gesloten, ben benieuwd hoe men er in de gebruikerstests op gaat reageren maar technisch is het zo in orde. Maar (ander issue) in dit accordion zijn het nog steeds links zonder aria-collapsed attributen, dat moet nog anders; zie https://designsystem.digital.gov/components/accordion/ voor een goed voorbeeld. Nieuwe issue aanmaken?

**Comment by PieterVerheij** [comment]
@timveld : gesloten als in okay bevonden?
